src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs

```diff
@@ -56,193 +56,104 @@ protected string CurrentUri
             }
         }
 
-        protected TOptions Options { get; private set; }
-
-        protected AuthenticateResult InitializeResult { get; private set; }
+        protected AuthenticationHandler(IOptionsSnapshot<TOptions> options, ILoggerFactory logger, UrlEncoder encoder, ISystemClock clock)
+        {
+            Logger = logger.CreateLogger(this.GetType().FullName);
+            UrlEncoder = encoder;
+            Clock = clock;
+            OptionsSnapshot = options;
+        }
 
         /// <summary>
-        /// Initialize is called once per request to contextualize this instance with appropriate state.
+        /// Initialize the handler, resolve the options and validate them.
         /// </summary>
-        /// <param name="options">The original options passed by the application control behavior</param>
-        /// <param name="context">The utility object to observe the current request and response</param>
-        /// <param name="logger">The logging factory used to create loggers</param>
-        /// <param name="encoder">The <see cref="UrlEncoder"/>.</param>
-        /// <returns>async completion</returns>
-        public async Task InitializeAsync(TOptions options, HttpContext context, ILogger logger, UrlEncoder encoder)
+        /// <param name="scheme"></param>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public virtual Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
         {
-            if (options == null)
+            if (scheme == null)
             {
-                throw new ArgumentNullException(nameof(options));
+                throw new ArgumentNullException(nameof(scheme));
             }
-
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            if (logger == null)
-            {
-                throw new ArgumentNullException(nameof(logger));
-            }
-
-            if (encoder == null)
-            {
-                throw new ArgumentNullException(nameof(encoder));
-            }
-
-            Options = options;
+            Scheme = scheme;
             Context = context;
-            OriginalPathBase = Request.PathBase;
-            OriginalPath = Request.Path;
-            Logger = logger;
-            UrlEncoder = encoder;
-
-            RegisterAuthenticationHandler();
-
-            Response.OnStarting(OnStartingCallback, this);
 
-            if (ShouldHandleScheme(AuthenticationManager.AutomaticScheme, Options.AutomaticAuthenticate))
+            Options = OptionsSnapshot.Get(Scheme.Name) ?? new TOptions();
```

Comment:
So its pretty much an error in any implementation of IOptionsSnapshot for Get to return null, so we could actually just assume that it always returns not null.